### PR TITLE
Sort aggregate report table rows hierarchically by state, district, s…

### PR DESCRIPTION
**Pull request by wtritch**
…chool

This updates our aggregate report result viewer to sort the organization column by:
```
State
District X
School X_A
School X_B
District M
School M_A
School M_B
```

Notice I used X and M because while State will always be first and Schools will be alpha-sorted by name under their districts, (District+School) groups will be sorted by district id.
This is a limitation of Schools not having access to their district names.  To correct the issue, we'd need to scan the (potentially thousands of) result rows to build a districtId->districtName map to query against during sorting.  If we expect to have human-readable multi-district reports that the user expects to read in alphabetical district order I can put it in.

Thoughts?

### CODE REVIEW

File: webapp/src/main/webapp/src/app/aggregate-report/results/aggregate-report-table.component.ts [hunk 1]
```diff
@@ -17,6 +18,29 @@ const OrderingDimensionType: Ordering<AggregateReportItem> = ordering(
 const OrderingDimensionValue: Ordering<AggregateReportItem> = ordering(byString)
   .on((item) => item.dimensionValue.toString());
 
+const OrderingOrganizationState: Ordering<AggregateReportItem> = ordering(
+  ranking([OrganizationType.State])
+).on((item) => item.organization.type);
```
Comment (wtritch) [inline webapp/src/main/webapp/src/app/aggregate-report/results/aggregate-report-table.component.ts]: This pulls State rows to the top of the list.

File: webapp/src/main/webapp/src/app/aggregate-report/results/aggregate-report-table.component.ts [hunk 2]
```diff
@@ -17,6 +18,29 @@ const OrderingDimensionType: Ordering<AggregateReportItem> = ordering(
...
+  .on((item) => {
+    switch(item.organization.type) {
+      case OrganizationType.District:
+        return (item.organization as District).id;
+      case OrganizationType.School:
+        return (item.organization as School).districtId;
+      default:
+        return -1;
+    }
+});
```
Comment (wtritch) [inline webapp/src/main/webapp/src/app/aggregate-report/results/aggregate-report-table.component.ts]: This bundles school rows with their district rows.

File: webapp/src/main/webapp/src/app/aggregate-report/results/aggregate-report-table.component.ts [hunk 3]
```diff
@@ -17,6 +18,29 @@ const OrderingDimensionType: Ordering<AggregateReportItem> = ordering(
...
+
+const OrderingOrganizationDistrict: Ordering<AggregateReportItem> = ordering(
+  ranking([OrganizationType.District])
+).on((item) => item.organization.type);
```
Comment (wtritch) [inline webapp/src/main/webapp/src/app/aggregate-report/results/aggregate-report-table.component.ts]: This pulls district rows to the top of their (district+school) row groups.

File: webapp/src/main/webapp/src/app/aggregate-report/results/aggregate-report-table.component.ts [hunk 4]
```diff
@@ -17,6 +18,29 @@ const OrderingDimensionType: Ordering<AggregateReportItem> = ordering(
...
+
+const OrderingOrganizationSchools: Ordering<AggregateReportItem> = ordering(byString)
+  .on((item) => item.organization.name);
```
Comment (wtritch) [inline webapp/src/main/webapp/src/app/aggregate-report/results/aggregate-report-table.component.ts]: This alpha-sorts schools by name within their district-scoped sub-groups.

File: webapp/src/main/webapp/src/app/aggregate-report/results/aggregate-report-table.component.ts [hunk 5]
```diff
@@ -73,7 +97,12 @@ export class AggregateReportTableComponent implements OnInit {
   private previousSort: any;
 
   constructor(public colorService: ColorService) {
-    this.orderingByProperty.organization = ordering(byString).on((item) => item.organizationName);
+    this.orderingByProperty.organization = ordering(join.apply(null, [
+      OrderingOrganizationState.compare,
+      OrderingOrganizationDistrictsWithSchools.compare,
+      OrderingOrganizationDistrict.compare,
+      OrderingOrganizationSchools.compare
+    ]));
```
Comment (wtritch) [inline webapp/src/main/webapp/src/app/aggregate-report/results/aggregate-report-table.component.ts]: This puts together the organization comparators such that they fall through to each successive comparator when the previous comparator returns an "equal" value.

File: webapp/src/main/webapp/src/app/aggregate-report/results/mock-aggregate-reports-preview.service.ts [hunk 6]
```diff
@@ -80,8 +82,7 @@ export class MockAggregateReportsPreviewService {
     uiModel.assessmentId = apiModel.assessment.id;
     uiModel.gradeId = apiModel.assessment.gradeId;
     uiModel.schoolYear = apiModel.examSchoolYear;
-    uiModel.organizationType = apiModel.organization.type;
-    uiModel.organizationName = apiModel.organization.name;
+    uiModel.organization = this.organizationMapper.map(apiModel.organization);
```
Comment (wtritch) [inline webapp/src/main/webapp/src/app/aggregate-report/results/mock-aggregate-reports-preview.service.ts]: I removed the flattened organization properties on the report items in favor of an Organization property so I could leverage the existing common OrganizationMapper.

File: webapp/src/main/webapp/src/app/aggregate-report/aggregate-reports-preview-table.component.ts [hunk 7]
```diff
@@ -73,7 +73,7 @@ export class AggregateReportsPreviewTableComponent implements OnInit {
 
   constructor(public colorService: ColorService,
               private assessmentDefinitionService: AssessmentDefinitionService) {
-    this.orderingByProperty.organizationName = ordering(byString).on((item) => item.organizationName);
+    this.orderingByProperty.organizationName = ordering(byString).on((item) => item.organization.name);
```
Comment (wtritch) [inline webapp/src/main/webapp/src/app/aggregate-report/aggregate-reports-preview-table.component.ts]: It concerns me that we have an aggregate-reports-preview-table.component.ts that's nearly a carbon-copy of the aggregate-report-table.component.ts ... This is going to lead to inconsistency errors...

File: webapp/src/main/webapp/src/app/aggregate-report/results/aggregate-report-table.component.ts (hunk 5 above)
Comment (esotrope) [reply webapp/src/main/webapp/src/app/aggregate-report/results/aggregate-report-table.component.ts]:
```javascript
fn.apply(this, [a, b, c]) === this.fn(a, b, c) 
```

Comment (wtritch) [reply webapp/src/main/webapp/src/app/aggregate-report/results/aggregate-report-table.component.ts]: um... good point.  Not sure why I did that.

File: webapp/src/main/webapp/src/app/aggregate-report/results/mock-aggregate-reports-preview.service.ts (hunk 6 above)
Comment (esotrope) [inline webapp/src/main/webapp/src/app/aggregate-report/results/mock-aggregate-reports-preview.service.ts]: yay reuse

File: webapp/src/main/webapp/src/app/aggregate-report/results/aggregate-report-table.component.ts (hunk 5 above)
Comment (esotrope) [reply webapp/src/main/webapp/src/app/aggregate-report/results/aggregate-report-table.component.ts]: you just wanted to do it the fun way i guess : j

File: webapp/src/main/webapp/src/app/aggregate-report/aggregate-reports-preview-table.component.ts (hunk 7 above)
Comment (esotrope) [reply webapp/src/main/webapp/src/app/aggregate-report/aggregate-reports-preview-table.component.ts]: After updating both in the last PR i share the same sentiment

**Comment by jtreuting** [comment]
What happens when there are just districts and no schools?

**Comment by wtritch** [comment]
@jtreuting The state will be first, and the districts will be sorted by DB id =/

**Comment by esotrope** [comment]
if we wanted to go by district name the backend will have to resolve the district name for each school. like custom org export we cant do it on the front end because the user may not be entitled to the district the school references.

**Comment by wtritch** [comment]
I don't think we'd need to resolve it on the backend because if the district row isn't visible, it doesn't matter what the order we show the grouped-by-district schools in.

**Comment by esotrope** [comment]
@wtritch you're right - i forgot this is in the results view when the rows contain the districts and stuff already 👍

**Comment by wtritch** [comment]
@jtreuting I added sorting (district+school) groups by district name.  I left in sorting (district+school) groups by district id to handle the unlikely case where two districts have the same name.

**Pull request merged**